### PR TITLE
IGNITE-22061 Use compact representation of BigDecimal to improve values size estimation

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-22061

**Issue description:**
Binary tuple stores bigdecimal in compact form after [IGNITE-21745](https://issues.apache.org/jira/browse/IGNITE-21745).

But `TupleMarshallerImpl.gatherStatistics()` and `ObjectStatistics.collectObjectStats()` estimate the size of `BigDecimal`s, without compaction, which can cause the pre-allocated buffer in the `BinaryTupleBuilder` to be significantly larger than required.

**Implementation notes:**
* Added conversion to compact form when collecting statistics for BigDecimals in `TupleMarshallerImpl.gatherStatistics()` and `ObjectStatistics.collectObjectStats()`.
* Marshallers have been slightly reworked to store values ​​obtained when collecting statistics.

### CODE REVIEW

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/ObjectStatistics.java [hunk 1]
```diff
@@ -23,35 +23,81 @@
...
+    int estimatedValueSize() {
         return estimatedValueSize;
     }
 
+    @TestOnly
+    <T> T value(int index) {
+        return (T) values[index];
+    }
+
+    static MarshallerRowBuilder createRowBuilder(SchemaDescriptor schema, Marshaller keyMarsh, Object key) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/ObjectStatistics.java:56]: Expression like `ObjectStatistics.createRowBuilder(...)` looks weird. 

Let's move factory methods to MarshallerRowBuilder class or MarshallerUtil.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/MarshallerUtil.java [hunk 2]
```diff
@@ -69,6 +70,23 @@ public static int getValueSize(Object val, NativeType type) throws InvalidTypeEx
...
     }
 
+    /**
+     * Converts the passed value to a more compact form, if possible.
+     *
+     * @param value Field value.
+     * @param type Mapped type.
+     * @return Value in a more compact form, or the original value if it cannot be compacted.
+     */
+    public static Object shrinkValue(Object value, NativeType type) {
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/MarshallerUtil.java:80]:
```suggestion
    public static <T> T shrinkValue(T value, NativeType type) {
```

File: modules/table/src/main/java/org/apache/ignite/internal/schema/marshaller/TupleMarshallerImpl.java [hunk 3]
```diff
@@ -187,15 +188,18 @@ private void gatherStatistics(
             }
 
             col.validate(val);
-            targetTuple.values.put(col.name(), val);
 
             if (val != null) {
+                val = MarshallerUtil.shrinkValue(val, col.type());
```
Comment (AMashenkov) [inline modules/table/src/main/java/org/apache/ignite/internal/schema/marshaller/TupleMarshallerImpl.java:193]: Does ti make sense for fix-length types (true == colType.spec().fixedLength()) ?

Comment (xtern) [reply modules/table/src/main/java/org/apache/ignite/internal/schema/marshaller/TupleMarshallerImpl.java]: Done, thanks

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/MarshallerUtil.java (hunk 2 above)
Comment (xtern) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/MarshallerUtil.java]: Done, thanks

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/ObjectStatistics.java (hunk 1 above)
Comment (xtern) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/reflection/ObjectStatistics.java]: Moved to `MarshallerRowBuilder`, thanks.

**Pull request closed without merging**